src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+import           Data.List.Split (chunksOf)
+import qualified Data.Map.Strict as M
+import qualified Data.Set as Set
+import           GHC.Natural
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+type ActivePeer c  = NodeInfo c
+type ActivePeers c = HashSet (ActivePeer c)
+
+data SyncEvent c tx s =
+      SyncBlock (Block c tx (Sealed c s))
+    | SyncBlockHeader (BlockHeader c s)
+
+deriving instance ( Eq (Hash c)
+                  , Eq tx
+                  , Eq s
+                  , Eq (PublicKey c)
+                  ) => Eq (SyncEvent c tx s)
+deriving instance ( Ord (Hash c)
+                  , Ord tx
+                  , Ord s
+                  , Ord (PublicKey c)
+                  ) => Ord (SyncEvent c tx s)
+deriving instance ( Show (Hash c)
+                  , Show tx
+                  , Show s
+                  , Show (PublicKey c)
+                  ) => Show (SyncEvent c tx s)
+
+data SyncError =
+      RequestTimeout ProtocolRequest Addr Timeout
+      -- ^ The given peer exceeded the request timeout when serving this
+      -- 'SyncProtocolRequest'.
+    | NoActivePeers
+      -- ^ There are not active peers to talk to.
+    | NoRemoteTipFound
+      -- ^ It was not possible to fetch a valid tip from our peers.
+    | AllPeersTimeoutError
+      -- ^ All the queried peers didn't respond in the allocated timeout.
+    deriving (Show, Eq)
+
+instance Exception SyncError
+
+data Timeout =
+    MaxTimeoutExceeded (Expected Duration)
+    -- ^ The operation exceeded the expected timeout, in nanoseconds.
+    deriving (Show, Eq)
+
+-- | A closed (inclusive) range [lo,hi].
+data Range = Range
+    { start :: Height
+    , end   :: Height
+    } deriving Show
+
+data ProtocolRequest =
+      GetTip
+    -- ^ Get the tip of the best chain.
+    | GetBlocks
+    -- ^ Get some blocks.
+    | GetBlockHeaders
+    -- ^ Get some block headers.
+    deriving (Show, Eq)
+
+-- | The singleton type for 'ProtocolRequest', capable of carrying some
+-- witness (in this case, 'ProtocolRequest').
+data SProtocolRequest r where
+  SGetTip :: SProtocolRequest 'GetTip
+  SGetBlocks :: SProtocolRequest 'GetBlocks
+  SGetBlockHeaders :: SProtocolRequest 'GetBlockHeaders
+
+-- | A closed type family mapping each 'ProtocolRequest' to its list of
+-- arguments, so that we can use the 'ProtocolRequest' (where all constructors
+-- have kind /*/) as a constraint.
+type family ProtocolRequestArgs (c :: ProtocolRequest) :: * where
+    ProtocolRequestArgs 'GetTip          = ()
+    ProtocolRequestArgs 'GetBlocks       = Range
+    ProtocolRequestArgs 'GetBlockHeaders = Range
+
+-- | A protocol response, indexed by the corresponding 'ProtocolRequest, so
+-- that matching the wrong type will result in a type error.
+type family ProtocolResponse c tx s (r :: ProtocolRequest) :: *
+
+type instance ProtocolResponse c (Tx c) PoW 'GetTip =
```

Comment:
Is there a reason the sync code can't be polymorphic in `tx`?

Reply:
And the seal?

Reply:
I was originally afraid that if I wasn't monomorphic in the `tx` and `s`, the compiler wouldn't have been able to specialise the typefam and actually apply. However, in the last version of my code `tx` is completely generic, so I might perhaps take another stab with `PoW` and see how far I get.